internal/controller/maintenancemode_controller.go

```diff
@@ -0,0 +1,188 @@
+			return !reflect.DeepEqual(oldObj.Status.InMaintenanceMode, newObj.Status.InMaintenanceMode)
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).Named("Maintenance Mode").
+		Watches(
+			&ramenv1alpha1.MaintenanceMode{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate),
+		).
+		Watches(
+			&v1alpha1.StorageClaim{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate),
+		).
+		Watches(
+			&v1alpha1.StorageClient{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate, storageClientStatusPredicate),
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes,verbs=get;list;update;create;watch;delete
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclients;storageclaims,verbs=get;list;watch
+
+func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "MaintenanceMode", req)
+	r.log.Info("Reconciling MaintenanceMode")
+
+	mapClientNameToObj := map[string]*v1alpha1.StorageClient{}
+
+	maintenanceModes := &ramenv1alpha1.MaintenanceModeList{}
+	if err := r.list(maintenanceModes); err != nil {
+		r.log.Error(err, "failed to list the Maintenance Modes")
+		return reconcile.Result{}, err
+	}
+
+	for i := range maintenanceModes.Items {
+		mm := &maintenanceModes.Items[i]
+		sc := &v1alpha1.StorageClaim{}
+		// MMode's TargetID is replicationID, which in our case is storageClaim name
+		sc.Name = mm.Spec.TargetID
+		err := r.get(sc)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		clientName := sc.Spec.StorageClient
+		if mapClientNameToObj[clientName] == nil {
+			storageClient := &v1alpha1.StorageClient{}
+			storageClient.Name = clientName
+			err := r.get(storageClient)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			mapClientNameToObj[clientName] = storageClient
+		}
+		if mapClientNameToObj[clientName].Status.InMaintenanceMode {
+			err := r.updateStatusCompletedForMM(mm)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	storageClients := &v1alpha1.StorageClientList{}
+	if err := r.list(storageClients); err != nil {
+		r.log.Error(err, "failed to list the Storage Clients")
+		return reconcile.Result{}, err
+	}
+
+	for i := range storageClients.Items {
+		storageClient := &storageClients.Items[i]
+		if mapClientNameToObj[storageClient.Name] != nil && !storageClient.Status.InMaintenanceMode {
+			if err := r.sendRequest(storageClient, true); err != nil {
+				return ctrl.Result{}, err
+			}
+		} else if mapClientNameToObj[storageClient.Name] == nil && storageClient.Status.InMaintenanceMode {
+			if err := r.sendRequest(storageClient, false); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
You don't need to separate cases here. Just check if the desired state (`needsMaintenanceMode`) is different from the actual state (`.Status.InMaintenanceMode`)
```suggestion
		_ , needsMaintenanceMode  = mapClientNameToObl[storageClient.Name]
		if needsMaintenanceMode != storageClient.Status.InMaintenanceMode {
			if err := r.sendRequest(storageClient, clientRequestMM); err != nil {
				return ctrl.Result{}, err
			}
		}
```